src/driver/mod.rs

```diff
@@ -165,14 +230,139 @@ impl<E: Engine> Driver<E> {
     async fn advance(&mut self) -> Result<()> {
         self.advance_safe_head().await?;
         self.advance_unsafe_head().await?;
+        self.run_sequencer_step().await?;
 
         self.update_finalized();
+        self.update_sync_status().await?;
         self.update_metrics();
+
         self.try_start_networking()?;
 
         Ok(())
     }
 
+    /// Prepare data for generating next payload.
+    fn prepare_block_data(
+        &self,
+        unsafe_epoch: Epoch,
+        new_blocktime: u64,
+    ) -> Result<(Epoch, L1Info)> {
+        let state = self.state.read().expect("lock poisoned");
+
+        // Check we are in sync with L1.
+        let current_epoch = state
+            .epoch_by_number(unsafe_epoch.number)
+            .ok_or(SequencerErr::OutOfSyncL1)?;
+
+        // Check past sequencer drift.
+        let is_seq_drift = new_blocktime > current_epoch.timestamp + self.max_seq_drift;
+        let next_epoch = state.epoch_by_number(current_epoch.number + 1);
+
+        let origin_epoch = if let Some(next_epoch) = next_epoch {
+            if new_blocktime >= next_epoch.timestamp {
+                next_epoch
+            } else {
+                current_epoch
+            }
+        } else {
+            if is_seq_drift {
+                return Err(SequencerErr::PastSeqDrift.into());
+            }
+            return Err(SequencerErr::NoNextEpoch.into());
+        };
+
+        let l1_info = state
+            .l1_info_by_hash(origin_epoch.hash)
+            .ok_or(SequencerErr::Critical(
+                "can't find l1 info for origin epoch during block building".to_string(),
+            ))?;
+
+        Ok((origin_epoch, l1_info.clone()))
+    }
+
+    /// Runs the sequencer step.
+    /// Produces a block if the conditions are met.
+    /// If successful the block would be signed by sequencer and shared by P2P.
+    async fn run_sequencer_step(&mut self) -> Result<()> {
+        if let Some(seq_config) = self.sequencer_config.clone() {
+            // Get unsafe head to build a new block on top of it.
+            let head = self.engine_driver.unsafe_info.head;
+            let unsafe_epoch = self.engine_driver.unsafe_info.epoch;
+
+            if seq_config.max_safe_lag() > 0 {
```

Comment:
I'm not sure if this extra check here really matters in terms of efficiency. Might be better to just remove it for readability reasons.

Reply:
I don't think that we need to remove the parameter, but doesn't the check on L294 cover this case, making it redundant, or am I misunderstanding something?

Reply:
Oh I see that makes sense!